packages/zipkin/src/record.js

```diff
@@ -0,0 +1,63 @@
+const {now, hrtime} = require('./time');
+const {Some, None} = require('./option');
+
+function Endpoint({serviceName, host, port}) {
+  this.serviceName = serviceName;
+  this.host = host || 0;
+  this.port = port || 0;
+}
+Endpoint.prototype.isUnknown = function isUnknown() {
+  return this.serviceName === undefined || this.host === 0 && this.port === 0;
+};
+
+function Annotation(timestamp, value) {
+  this.timestamp = timestamp;
+  this.value = value;
+}
+Annotation.prototype.toString = function toString() {
+  return `Annotation(value="${this.value}")`;
+};
+
+function Span(traceId) {
+  this.traceId = traceId;
+  this.startTimestamp = now();
+  this.startTick = hrtime();
+  this.name = None;
+  this.service = None;
+  this.endpoint = new Endpoint({});
+  this.remoteEndpoint = None;
+  this.annotations = [];
+  this.tags = {};
+}
+Span.prototype.setName = function setName(name) {
+  this.name = new Some(name);
+};
+Span.prototype.setServiceName = function setServiceName(name) {
+  this.service = new Some(name);
+};
+Span.prototype.setRemoteEndpoint = function setRemoteEndpoint(ep) {
+  this.remoteEndpoint = new Some(ep);
+};
+Span.prototype.addAnnotation = function addAnnotation(timestamp, value) {
+  if (!this.endTimestamp && (
+        value === 'cr' ||
+        value === 'ss'
+        )) {
+    this.endTimestamp = now(this.startTimestamp, this.startTick);
+  }
+  this.annotations.push(new Annotation(timestamp, value));
+};
+Span.prototype.putTag = function putTag(key, value) {
+  this.tags[key] = value;
+};
+Span.prototype.setEndpoint = function setEndpoint(ep) {
+  this.endpoint = ep;
+};
+Span.prototype.toString = function toString() {
+  const annotations = this.annotations.map(a => a.toString()).join(', ');
+  return `Span(id=${this.traceId.toString()}, annotations=[${annotations}])`;
+};
+
+module.exports.Endpoint = Endpoint;
+module.exports.Annotation = Annotation;
+module.exports.Span = Span;
```

Comment:
fyi @narayaruna this will make the "Span" type visible for future work, such as fancier batching. This type is almost v2 now, but I don't want to do the full v2 conversion in this pull request, as this one is about making thrift completely optional